[apr] fix CMake config via patch from the stable branch.

rinrab opened the pull request:
Some of the changes made in patch which fixes CMake config in the apr port were moved to the upstream. I added these changes the port via patch, made from upstream.

* Use patch with changes from the stable branch to fix CMake config instead of the fix-configcmake.patch file; Remove yhis file.
* Take windows part of the portfile.cmake file from stable branch of apr.
* Change CMake namespace from `unofficial-apr` to just `apr`, patch is now taken from the stable branch of apr.

rinrab committed:
[apr] use patch and portfile from the stable branch of apr.

* Use patch with changes from the stable branch to fix cmake config instead of fix-configcmake.patch file.
* Take windows part of the portfile.cmake file from stable branch of apr.
* Remove the fix-configcmake.patch file.
* Change cmake namespace from `unofficial-apr` to just `apr`, patch is now taken from the stable branch of apr.

rinrab committed:
[apr] remove the `unglue.patch` file.

rinrab committed:
[apr] add new version of the apr port.

rinrab committed:
[apr] add the port-version property to the vcpkg.json file.

rinrab committed:
[apr] update git-tree.

rinrab committed:
[apr] remove setting of the APU_USE_EXPAT variable; It is never used.

rinrab committed:
[apr] add the license field to the vcpkg.json file.

rinrab committed:
[apr] specify the `default-features` field to the vcpkg.json file.

rinrab committed:
[apr] run the `.\vcpkg.exe format-manifest` command.

rinrab committed:
[apr] update git-tree.

rinrab committed:
Revert "[apr] remove the `unglue.patch` file."

This reverts commit 430a1faf430f8404b50f3cdd56b8710d27df6208.

rinrab committed:
[apr] update git-tree.

rinrab committed:
[apr] add the license field to the vcpkg.json file.

rinrab committed:
[apr] update git-tree.

WangWeiLin-MV commented:
Local attempt to apply upstream-diff, but failed.
````cmake
vcpkg_download_distfile(PATCH_APR_1917612_FROM_1909837
    URLS "https://svn.apache.org/viewvc/apr/apr/branches/1.7.x/CMakeLists.txt?r1=1917612&r2=1909837&view=patch"
    FILENAME "PATCH_APR_1917612_FROM_1909837.diff"
    SHA512 4bd8943d224bf90ec215c3f09d3c1ca44ba732db245ca2eb65092a71d1ce3b7aa9b444654c155857a8bfd26adb1934af2be1563bd4317518ebdceab0f816709b
)
file(COPY_FILE "${DOWNLOADS}/PATCH_APR_1917612_FROM_1909837.diff" "${CURRENT_BUILDTREES_DIR}/patches/PATCH_APR_1917612_FROM_1909837.diff"}
vcpkg_replace_string("${CURRENT_BUILDTREES_DIR}/patches/PATCH_APR_1917612_FROM_1909837.diff"
    "apr/apr/branches/1.7.x/CMakeLists.txt"
    "CMakeLists.txt"
)

vcpkg_extract_source_archive(SOURCE_PATH
    ARCHIVE "${ARCHIVE}"
    PATCHES
        "${CURRENT_BUILDTREES_DIR}/patches/PATCH_APR_1917612_FROM_1909837.diff"
        unglue.patch
)
````

WangWeiLin-MV reviewed (dismissed):
Patch diffs are basically the same, and can be viewed via compare the pull-diff
https://github.com/microsoft/vcpkg/blob/fa826f088dded4b387a503ee6303ced0acbc7363/ports/apr/apr-r1917051%2Br1917283%2Br1917612.patch
and upstream-diff
https://svn.apache.org/viewvc/apr/apr/branches/1.7.x/CMakeLists.txt?r1=1917612&r2=1909837&view=patch

---

Edit: Newer changes, test invalid
~The port usage tests pass with the following triplets:~
* ~x64-windows~

rinrab commented:
I checked this problem and I found that the patch works as well.

I did the following:

1. Checkout the tag of the repository from https://svn.apache.org/repos/asf/apr/apr/tags/1.7.4/
2. Apply the patch via svn: `svn patch path\to\patch\apr-r1917051+r1917283+r1917612.patch`

I also tried to do it via git: `cat path\to\patch\apr-r1917051+r1917283+r1917612.patch | git apply` and it worked.

WangWeiLin-MV commented:
@rinrab Sorry for not clarify, the patches in this PR can be applied. My local attempt is using the patch downloaded from the [link mentioned](https://github.com/microsoft/vcpkg/pull/38674#pullrequestreview-2051077356).

If possible, we prefer patches by downloading upstream patches [like this](https://github.com/microsoft/vcpkg/pull/38674#issuecomment-2105555119) rather than add patches to the port.

rinrab committed:
[apr] generate patches per all revisions

Patches are generated via the following command:
```powershell
1917051,1917065,1917283,1917612 | ForEach-Object { svn-diff https://svn.apache.org/repos/asf/apr/apr/branches/1.7.x -c $_ -raw | Out-File "apr-r$_.patch" }
```

rinrab committed:
[apr] fix patch appling

* apr-r1917051.patch: Replace ENDIF() with SET(install_bin_pdb) in context.
* apr-r1917283.patch: Remove changes in the CHANGES file because it doesn't affect on the result.

rinrab committed:
[apr] update git-tree

rinrab commented:
I don't think that this solution is good because:

1. viewvc is not an official apache resource. The only resources that are official are downloads.apache.org, archive.apache.org, dlcdn.apache.org, and svn.apache.org over the Subversion protocols.
2. The diff is being regenerated on every request, so this is not a static content.
3. Diff format which is used could be changed.

For better clarity I separated that patch by each revisions.

rinrab commented:
> Local attempt to apply upstream-diff, but failed.
> 
> ```cmake
> vcpkg_download_distfile(PATCH_APR_1917612_FROM_1909837
>     URLS "https://svn.apache.org/viewvc/apr/apr/branches/1.7.x/CMakeLists.txt?r1=1917612&r2=1909837&view=patch"
>     FILENAME "PATCH_APR_1917612_FROM_1909837.diff"
>     SHA512 4bd8943d224bf90ec215c3f09d3c1ca44ba732db245ca2eb65092a71d1ce3b7aa9b444654c155857a8bfd26adb1934af2be1563bd4317518ebdceab0f816709b
> )
> file(COPY_FILE "${PATCH_APR_1917612_FROM_1909837}" "${CURRENT_BUILDTREES_DIR}/patches/PATCH_APR_1917612_FROM_1909837.diff"}
> vcpkg_replace_string("${CURRENT_BUILDTREES_DIR}/patches/PATCH_APR_1917612_FROM_1909837.diff"
>     "apr/apr/branches/1.7.x/CMakeLists.txt"
>     "CMakeLists.txt"
> )
> 
> vcpkg_extract_source_archive(SOURCE_PATH
>     ARCHIVE "${ARCHIVE}"
>     PATCHES
>         "${CURRENT_BUILDTREES_DIR}/patches/PATCH_APR_1917612_FROM_1909837.diff"
>         unglue.patch
> )
> ```
> 
> @rinrab Could you help apply patches by this method?

Your attempt failed because the context in patch was wrong.

When patch is applying, it looks at the lines before and after. However they were changed in the '1.7.x' branch before these changes. It's required to fix them like I did here: https://github.com/microsoft/vcpkg/pull/38674/commits/f23193dc7bcc1403018d87fda81391357216943c

rinrab commented:
> Local attempt to apply upstream-diff, but failed.
> 
> ```cmake
> vcpkg_download_distfile(PATCH_APR_1917612_FROM_1909837
>     URLS "https://svn.apache.org/viewvc/apr/apr/branches/1.7.x/CMakeLists.txt?r1=1917612&r2=1909837&view=patch"
>     FILENAME "PATCH_APR_1917612_FROM_1909837.diff"
>     SHA512 4bd8943d224bf90ec215c3f09d3c1ca44ba732db245ca2eb65092a71d1ce3b7aa9b444654c155857a8bfd26adb1934af2be1563bd4317518ebdceab0f816709b
> )
> file(COPY_FILE "${PATCH_APR_1917612_FROM_1909837}" "${CURRENT_BUILDTREES_DIR}/patches/PATCH_APR_1917612_FROM_1909837.diff"}
> vcpkg_replace_string("${CURRENT_BUILDTREES_DIR}/patches/PATCH_APR_1917612_FROM_1909837.diff"
>     "apr/apr/branches/1.7.x/CMakeLists.txt"
>     "CMakeLists.txt"
> )
> 
> vcpkg_extract_source_archive(SOURCE_PATH
>     ARCHIVE "${ARCHIVE}"
>     PATCHES
>         "${CURRENT_BUILDTREES_DIR}/patches/PATCH_APR_1917612_FROM_1909837.diff"
>         unglue.patch
> )
> ```
> 
> @rinrab Could you help apply patches by this method?


Your attempt failed because the context in patch was wrong.

When patch is applying, it looks at the lines before and after. However they were changed in the '1.7.x' branch before these changes. It's required to fix them like I did here: https://github.com/microsoft/vcpkg/pull/38674/commits/f23193dc7bcc1403018d87fda81391357216943c (In github webiu it looks so strange. I don't know why)

WangWeiLin-MV commented on ports/apr/portfile.cmake:
```diff
@@ -10,31 +10,38 @@ vcpkg_download_distfile(ARCHIVE
     vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
-        FEATURES
-            private-headers APR_INSTALL_PRIVATE_H
+    FEATURES
+        private-headers INSTALL_PRIVATE_H
```
Revert changes here
```suggestion
        FEATURES
            private-headers APR_INSTALL_PRIVATE_H
```

rinrab replied:
Sorry, my bad. Commit this change.

WangWeiLin-MV commented on ports/apr/portfile.cmake:
```diff
@@ -10,31 +10,38 @@ vcpkg_download_distfile(ARCHIVE
-            -DMIN_WINDOWS_VER=Windows7
-            -DAPR_HAVE_IPV6=ON
-            ${FEATURE_OPTIONS}
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS
+        -DAPR_BUILD_STATIC=${APR_BUILD_STATIC}
+        -DAPR_BUILD_SHARED=${APR_BUILD_SHARED}
+        -DAPR_BUILD_TESTAPR=OFF
+        -DINSTALL_PDB=OFF
+        -DAPR_INSTALL_PRIVATE_H=${INSTALL_PRIVATE_H}
```
Indent and revert `${FEATURE_OPTIONS}`
```suggestion
        SOURCE_PATH "${SOURCE_PATH}"
        OPTIONS
            -DAPR_BUILD_STATIC=${APR_BUILD_STATIC}
            -DAPR_BUILD_SHARED=${APR_BUILD_SHARED}
            -DAPR_BUILD_TESTAPR=OFF
            -DINSTALL_PDB=OFF
            ${FEATURE_OPTIONS}
```

rinrab replied:
I reverted accidentally changed indentation in https://github.com/microsoft/vcpkg/pull/38674/commits/faa5eb2e887cd0488e7f951099fc009b83baf1c9

About `${FEATURE_OPTIONS}` I disagree, because the feature variable name is `INSTALL_PRIVATE_H` while the cmake command `APR_INSTALL_PRIVATE_H`, so it would not work.

Additionally, I think that it is better to pass the options explicitly.

dg0yt replied:
`vcpkg_check_feautures` exist only to provide feature options to cmake. The original `APR_INSTALL_PRIVATE_H` / `FEATURE_OPTIONS` pattern is "canonical vcpkg". Please restore it.

rinrab committed:
[apr] revert accedently changed indentation in the portfile.cmake.

rinrab committed:
[apr] update git-tree

rinrab commented:
> Edit: Newer changes, test invalid 

I just checked, in my environment it does work in `x64-windows` and `x64-windows-static`.

WangWeiLin-MV commented on ports/apr/portfile.cmake:
```diff
@@ -10,31 +10,38 @@ vcpkg_download_distfile(ARCHIVE
     vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
         FEATURES
-            private-headers APR_INSTALL_PRIVATE_H
+            private-headers INSTALL_PRIVATE_H
```
```suggestion
            private-headers APR_INSTALL_PRIVATE_H
```

WangWeiLin-MV commented on ports/apr/portfile.cmake:
```diff
@@ -10,31 +10,38 @@ vcpkg_download_distfile(ARCHIVE
-            -DMIN_WINDOWS_VER=Windows7
-            -DAPR_HAVE_IPV6=ON
-            ${FEATURE_OPTIONS}
+            -DAPR_INSTALL_PRIVATE_H=${INSTALL_PRIVATE_H}
```
```suggestion
            ${FEATURE_OPTIONS}
```

rinrab replied:
Committed in https://github.com/microsoft/vcpkg/commit/e0e345420fc5fec27f0310ca55193601413f2d20

WangWeiLin-MV reviewed (changes requested):
@rinrab Therefore, it should change the variable corresponding to feature `private-headers` to `APR_INSTALL_PRIVATE_H`. Moreover, `FEATURE_OPTIONS` clearly pass the options and it is recorded in the features of `vcpkg.json`.

rinrab committed:
[apr] use the FEATURE_OPTIONS variable to pass features to the cmake configure command.

rinrab committed:
[apr] revert setting of the MIN_WINDOWS_VER and APR_HAVE_IPV6 variables.

rinrab committed:
[apr] update git-tree

rinrab commented:
> @rinrab Therefore, it should change the variable corresponding to feature `private-headers` to `APR_INSTALL_PRIVATE_H`. Moreover, `FEATURE_OPTIONS` clearly pass the options and it is recorded in the features of `vcpkg.json`.

I committed these suggestions in https://github.com/microsoft/vcpkg/pull/38674/commits/e0e345420fc5fec27f0310ca55193601413f2d20, however I prefer to follow the APR's upstream version of the portfile.

rinrab commented:
Those have been released as APR of version 1.7.5. I opened a pull-request with bumping the version and adding some features from this pull-request.
